File: services/research-backend/src/research_backend/infrastructure/messaging/contracts.py
"""RabbitMQ CollectionTask와 CollectionResult v1 Pydantic 계약을 정의한다."""

import hashlib
import json
from datetime import datetime
from typing import Annotated, Literal

from pydantic import (
    AwareDatetime,
    BaseModel,
    ConfigDict,
    Field,
    StringConstraints,
    model_validator,
)

from research_backend.clients.collector.models import (
    CollectorResult,
    MerchantName,
    SearchFilters,
)

QueueIdentifier = Annotated[
    str,
    StringConstraints(min_length=1, max_length=128, pattern=r"^[A-Za-z0-9][A-Za-z0-9._:-]*$"),
]
IdempotencyKey = Annotated[
    str,
    StringConstraints(pattern=r"^collection:v1:[a-f0-9]{64}$"),
]


class QueueModel(BaseModel):
    """Queue 계약의 JSON alias와 추가 필드 거부 규칙을 제공한다."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class SearchPayload(QueueModel):
    """Go 판매처 Adapter가 검색 요청을 만들 때 사용할 조건을 표현한다."""

    query: str = Field(min_length=1, max_length=200)
    page: int = Field(default=1, ge=1)
    limit: int = Field(default=3, ge=1, le=50)
    locale: str = Field(default="ko-KR", pattern=r"^[a-z]{2}-[A-Z]{2}$")
    currency: str = Field(default="KRW", pattern=r"^[A-Z]{3}$")
    filters: SearchFilters = Field(default_factory=SearchFilters)

    @model_validator(mode="after")
    def require_supported_page(self) -> "SearchPayload":
        """현재 Adapter가 지원하지 않는 2페이지 이상 작업을 등록 전에 거부한다."""

        if self.page != 1:
            raise ValueError("현재 검색 Worker는 page=1만 지원합니다")
        return self


class CollectionTask(QueueModel):
    """Python Backend가 RabbitMQ 검색 Queue에 등록하는 작업 봉투다."""

    schema_version: Literal["1"] = Field(default="1", alias="schemaVersion")
    task_id: QueueIdentifier = Field(alias="taskId")
    job_id: QueueIdentifier = Field(alias="jobId")
    merchant: MerchantName
    operation: Literal["search"] = "search"
    priority: int = Field(default=20, ge=0, le=100)
    attempt: int = Field(default=0, ge=0)
    max_attempts: int = Field(default=2, alias="maxAttempts", ge=1, le=5)
    requested_at: AwareDatetime = Field(alias="requestedAt")
    idempotency_key: IdempotencyKey = Field(alias="idempotencyKey")
    payload: SearchPayload

    @model_validator(mode="after")
    def validate_attempt(self) -> "CollectionTask":
        """현재 attempt가 최대 시도 횟수보다 작도록 보장한다."""

        if self.attempt >= self.max_attempts:
            raise ValueError("attempt는 maxAttempts보다 작아야 합니다")
        return self


class TaskError(QueueModel):
    """작업 실패 코드, 안전한 설명과 재시도 가능 여부를 표현한다."""

    code: str = Field(min_length=1, max_length=100, pattern=r"^[A-Z][A-Z0-9_]*$")
    message: str = Field(min_length=1, max_length=1000)
    retryable: bool


class CollectionResultEnvelope(QueueModel):
    """Go Worker가 결과 Queue에 발행하는 작업 실행 결과를 표현한다."""

    schema_version: Literal["1"] = Field(alias="schemaVersion")
    task_id: QueueIdentifier = Field(alias="taskId")
    job_id: QueueIdentifier = Field(alias="jobId")
    status: Literal["success", "partial", "failed"]
    started_at: AwareDatetime = Field(alias="startedAt")
    completed_at: AwareDatetime = Field(alias="completedAt")
    duration_ms: int = Field(alias="durationMs", ge=0)
    collector_result: CollectorResult | None = Field(alias="collectorResult")
    error: TaskError | None

    @model_validator(mode="after")
    def validate_status_payload(self) -> "CollectionResultEnvelope":
        """성공에는 CollectorResult만, 실패에는 TaskError가 반드시 있도록 검사한다."""

        if self.completed_at < self.started_at:
            raise ValueError("completedAt은 startedAt보다 빠를 수 없습니다")
        if self.status in {"success", "partial"}:
            if self.collector_result is None or self.error is not None:
                raise ValueError("성공·부분 성공 결과에는 collectorResult만 필요합니다")
            if self.collector_result.status != self.status:
                raise ValueError("Queue status와 Collector status가 일치해야 합니다")
        elif self.error is None:
            raise ValueError("실패 결과에는 error가 필요합니다")
        return self


def build_idempotency_key(
    merchant: str,
    operation: str,
    payload: SearchPayload,
) -> str:
    """판매처·작업·검색 조건을 정렬한 SHA-256 멱등성 키로 변환한다.

    동일 입력은 동일 키를 반환하며 민감정보나 임의 URL을 포함하지 않는다. 실제 중복
    등록 차단은 Redis adapter가 구현될 때 이 키를 사용한다.
    """

    canonical = json.dumps(
        {
            "merchant": merchant,
            "operation": operation,
            "payload": payload.model_dump(mode="json", by_alias=True, exclude_none=True),
        },
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    )
    digest = hashlib.sha256(canonical.encode("utf-8")).hexdigest()
    return f"collection:v1:{digest}"


def now_with_timezone() -> datetime:
    """Queue CLI가 timezone이 포함된 현재 시각을 만들 수 있도록 반환한다."""

    return datetime.now().astimezone()

File: services/collector/internal/messaging/contracts.go
// Package messaging은 RabbitMQ 작업·결과 계약과 Collector Worker 실행을 제공한다.
package messaging

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/leeseoin/opensource-competition/services/collector/internal/collector"
)

const (
	// SchemaVersion은 현재 Collection Queue 메시지 계약 버전이다.
	SchemaVersion = "1"
	// OperationSearch는 판매처 상품 검색 작업을 나타낸다.
	OperationSearch = "search"
	// TaskStatusSuccess는 Collector가 검색 범위를 정상 수집했음을 나타낸다.
	TaskStatusSuccess = "success"
	// TaskStatusPartial은 Collector가 검색 범위의 일부만 수집했음을 나타낸다.
	TaskStatusPartial = "partial"
	// TaskStatusFailed는 작업을 저장 가능한 결과로 완료하지 못했음을 나타낸다.
	TaskStatusFailed = "failed"
)

var (
	queueIdentifierPattern = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._:-]*$`)
	idempotencyKeyPattern  = regexp.MustCompile(`^collection:v1:[a-f0-9]{64}$`)
)

// CollectionTask는 Product Backend가 RabbitMQ에 등록하는 검색 작업 봉투다.
type CollectionTask struct {
	SchemaVersion  string        `json:"schemaVersion"`
	TaskID         string        `json:"taskId"`
	JobID          string        `json:"jobId"`
	Merchant       string        `json:"merchant"`
	Operation      string        `json:"operation"`
	Priority       int           `json:"priority"`
	Attempt        int           `json:"attempt"`
	MaxAttempts    int           `json:"maxAttempts"`
	RequestedAt    time.Time     `json:"requestedAt"`
	IdempotencyKey string        `json:"idempotencyKey"`
	Payload        SearchPayload `json:"payload"`
}

// SearchPayload는 판매처 Adapter가 검색 URL을 만들 때 사용할 Queue v1 조건이다.
// Adapter의 대량 수집용 SearchPage와 별개로 Queue v1은 Page가 1이 아니면 검증에 실패한다.
type SearchPayload struct {
	Query    string                  `json:"query"`
	Page     int                     `json:"page"`
	Limit    int                     `json:"limit"`
	Locale   string                  `json:"locale"`
	Currency string                  `json:"currency"`
	Filters  collector.SearchFilters `json:"filters"`
}

// TaskError는 작업 실패 코드, 안전한 설명과 재시도 가능 여부를 전달한다.
type TaskError struct {
	Code      string `json:"code"`
	Message   string `json:"message"`
	Retryable bool   `json:"retryable"`
}

// CollectionResultEnvelope는 Go Worker가 Product Backend 저장 Worker에 반환하는 실행 결과다.
type CollectionResultEnvelope struct {
	SchemaVersion   string                  `json:"schemaVersion"`
	TaskID          string                  `json:"taskId"`
	JobID           string                  `json:"jobId"`
	Status          string                  `json:"status"`
	StartedAt       time.Time               `json:"startedAt"`
	CompletedAt     time.Time               `json:"completedAt"`
	DurationMS      int64                   `json:"durationMs"`
	CollectorResult *collector.SearchResult `json:"collectorResult"`
	Error           *TaskError              `json:"error"`
}

// Validate는 작업 봉투와 검색 조건이 Collection Queue v1 계약을 만족하는지 검사한다.
func (t CollectionTask) Validate() error {
	if t.SchemaVersion != SchemaVersion {
		return fmt.Errorf("schemaVersion은 %s이어야 합니다", SchemaVersion)
	}
	if !validQueueIdentifier(t.TaskID) {
		return fmt.Errorf("taskId 형식이 올바르지 않습니다")
	}
	if !validQueueIdentifier(t.JobID) {
		return fmt.Errorf("jobId 형식이 올바르지 않습니다")
	}
	if t.Operation != OperationSearch {
		return fmt.Errorf("현재 operation은 search만 지원합니다")
	}
	if t.Priority < 0 || t.Priority > 100 {
		return fmt.Errorf("priority는 0 이상 100 이하여야 합니다")
	}
	if t.Attempt < 0 || t.MaxAttempts < 1 || t.MaxAttempts > 5 || t.Attempt >= t.MaxAttempts {
		return fmt.Errorf("attempt와 maxAttempts 범위가 올바르지 않습니다")
	}
	if t.RequestedAt.IsZero() {
		return fmt.Errorf("requestedAt이 필요합니다")
	}
	if !idempotencyKeyPattern.MatchString(t.IdempotencyKey) {
		return fmt.Errorf("idempotencyKey 형식이 올바르지 않습니다")
	}
	if t.Payload.Page != 1 {
		return fmt.Errorf("현재 Queue v1 검색 Worker는 page=1만 지원합니다")
	}
	return t.SearchRequest().Validate()
}

// SearchRequest는 Queue 작업을 기존 Collector 검색 요청으로 변환한다.
func (t CollectionTask) SearchRequest() collector.SearchRequest {
	return collector.SearchRequest{
		RequestID:   t.TaskID,
		Merchant:    t.Merchant,
		Query:       strings.TrimSpace(t.Payload.Query),
		RequestedAt: t.RequestedAt,
		Limit:       t.Payload.Limit,
		Locale:      t.Payload.Locale,
		Currency:    t.Payload.Currency,
		Filters:     t.Payload.Filters,
	}
}

// CanRetry는 현재 실행 실패 후 retry Queue로 보낼 수 있는지 반환한다.
func (t CollectionTask) CanRetry() bool {
	return t.Attempt+1 < t.MaxAttempts
}

// NextAttempt는 재시도 횟수를 하나 증가시킨 새 작업 값을 반환한다.
func (t CollectionTask) NextAttempt() CollectionTask {
	t.Attempt++
	return t
}

// validQueueIdentifier는 taskId와 jobId가 길이와 허용 문자 규칙을 만족하는지 확인한다.
func validQueueIdentifier(value string) bool {
	return len(value) >= 1 && len(value) <= 128 && queueIdentifierPattern.MatchString(value)
}
